### PR TITLE
feature: adds wait step creator for API Tests

### PR DESCRIPTION
**Pull request by rishiranjjan**
This MR adds support to create Wait steps for API Tests, which is currently available through the DD UI (_screenshot below_)

<kbd>
Screenshot 2024-11-04 at 18 20 08
</kbd>

### Testing

1. Ran `./gradlew e2eTest`
2. The test ran successfully and created the step on DD UI
<kbd>
Screenshot 2024-11-04 at 18 20 25
</kbd>

### CODE REVIEW

**Review by matt-long-92 — APPROVED** [review]
LGTM

File: src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt [hunk 1]
```diff
@@ -115,4 +132,15 @@ class StepBuilder(
         }
         retryOptions.count(retryCount).interval(retryInterval.inWholeMilliseconds.toDouble())
     }
+
+    /**
+     * Creates a wait step
+     * @param duration The duration to wait in seconds. Minimum value: 0. Maximum value: 180
+     */
+    fun wait(duration: Int) {
+        require(duration in 0..180) {
```
[the remark this answers is not in the document]
Comment (rishiranjjan) [reply src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt]: Yeah, 0 wait doesn't makes sense. I was kinda replicating DD's library from [here](https://github.com/DataDog/datadog-api-client-java/blob/92437a1ecc06ce26a5fc87e55f1e831cdc7aed5b/src/main/java/com/datadog/api/client/v1/model/SyntheticsAPIWaitStep.java#L103) but it is not possible from the UI to have 0 seconds wait duration (_obviously_). So changed to to `1 .. 180`

File: src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt [hunk 2]
```diff
@@ -115,4 +132,15 @@ class StepBuilder(
         }
         retryOptions.count(retryCount).interval(retryInterval.inWholeMilliseconds.toDouble())
     }
+
+    /**
+     * Creates a wait step
+     * @param durationInSeconds The duration to wait in seconds. Minimum value: 0. Maximum value: 180
```
Comment (kaivalyaapte) [inline src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt:138]: doc needs an update. it says 0

File: src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt [hunk 3]
```diff
@@ -32,30 +34,45 @@ class StepBuilder(
 
     private val step = SyntheticsAPIStep(SyntheticsAPITestStep())
     private var request: SyntheticsTestRequest? = null
+    private var waitDuration: Int? = null
```
[the remark this answers is not in the document]
Comment (kaivalyaapte) [reply src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt:37]: > Also kind of replicating the behaviour we have with request
This is reasonable and its fine to keep it, but honestly `null` checks should be avoided and we can make the logic simpler by using sensible defaults. Specifically for types like Ints, bools etc.

I would leave it up to you to decide.

File: src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt [hunk 4]
```diff
@@ -32,30 +34,45 @@ class StepBuilder(
...
             step.syntheticsAPITestStep.isCritical(isCritical)
         }
 
         if (parsingOptions.isNotEmpty()) {
             step.syntheticsAPITestStep.extractedValues(parsingOptions)
         }
+        return if (waitDuration != null) {
+            buildWaitStep()
+        } else {
+            SyntheticsAPIStep(
```
Comment (kaivalyaapte) [inline src/main/kotlin/com/personio/synthetics/builder/api/StepBuilder.kt:58]: nit: Can we also pull this in a function? Then we can simplify this if else with a ternary operator for example.

File: src/test/kotlin/com/personio/synthetics/builder/api/StepBuilderTest.kt [hunk 5]
```diff
@@ -69,6 +71,28 @@ class StepBuilderTest {
...
+            result.syntheticsAPIWaitStep,
+        )
+    }
+
+    @Test
+    fun `wait throws exception for invalid duration`() {
+        val stepBuilder = StepBuilder(TEST_STEP_NAME)
+
+        assertThrows<IllegalArgumentException> {
+            stepBuilder.wait(300)
```
Comment (kaivalyaapte) [inline src/test/kotlin/com/personio/synthetics/builder/api/StepBuilderTest.kt:92]: I would also test for the minimum value to make it a range check.

**Review by kaivalyaapte — APPROVED** [review]
Left some comments. But looks good. Approving.

**Review by kaivalyaapte — APPROVED** [review]
LGTM
